### PR TITLE
Add support for CUDA atomics using "unsigned long int"

### LINKED ISSUE #675: Allowed Types in CUDA atomic add
I get an compile error when using the CUDA atomic add function. If the CPU serial accelerator is used instead of the CUDA accelerator, it compiles and runs flawlessly. 

The code used to create this error can be found here: 
[atomicAddTest.cpp.txt](https://github.com/ComputationalRadiationPhysics/alpaka/files/2455412/atomicAddTest.cpp.txt)

The accompanying log can be found here:
[log.txt](https://github.com/ComputationalRadiationPhysics/alpaka/files/2455411/log.txt)

Edit:

Loaded Modules:
 * cmake 3.11.3
 * git 2.19.0
 * gcc 6.4.0
 * cuda 9.2
 * boost 1.68.0

Test System: 
 * 2x Intel(R) Xeon(R) Gold 6136 CPU @ 3.00GHz (12 Cores/24 Threads)
 * 4x NVIDIA Tesla P100
 * CentOS 7 (with Linux kernel 3.10)

### PR DESCRIPTION
**Pull request by BenjaminW3**
This fixes #675 and adds tests for more CUDA atomics and support for `unsigned long int` for CUDA atomics.
`unsigned long int` is equal to `unsigned int` on LLP64 platforms (Win64) and equal to `unsigned long long int` on LP64 platforms (unix).Depending on the platform the CUDA atomics called with `unsigned long int` support the operations of the matching type.

Now we have tests for CUDA atomics for the following types:
* unsigned int (was already available)
* int (new)
* unsigned long int (new)
* unsigned long long int (new)

### CODE REVIEW

**Comment by BenjaminW3** [comment]
Ready to be merged!

**Comment by ax3l** [comment]
Sorry for the delay, most of us were all week bound in organizing a on-site PIConGPU user workshop (successfully :) ) maybe @psychocoderHPC and @tdd11235813 can take a look?

**Comment by psychocoderHPC** [comment]
Yes I will have a look to ot tomorrow. I started the review today but get interupted in between.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 1]
```diff
@@ -41,294 +41,663 @@
...
+    {
+        testAtomicAdd(acc, success, operandOrig);
+        testAtomicSub(acc, success, operandOrig);
+
+        testAtomicMin(acc, success, operandOrig);
+        testAtomicMax(acc, success, operandOrig);
+
+        testAtomicExch(acc, success, operandOrig);
+
+        //testAtomicInc(acc, success, operandOrig);
```
Comment (psychocoderHPC) [inline test/unit/atomic/src/AtomicTest.cpp]: Why is this commented out? Please add a comment why it is not enabled or remove the code.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 2]
```diff
@@ -41,294 +41,663 @@
...
+    //-----------------------------------------------------------------------------
+    ALPAKA_NO_HOST_ACC_WARNING
+    ALPAKA_FN_ACC auto operator()(
+        alpaka::acc::AccGpuCudaRt<TDim, TIdx> const & acc,
+        bool * success,
+        unsigned long long int operandOrig) const
+    -> void
+    {
+        testAtomicAdd(acc, success, operandOrig);
+        //testAtomicSub(acc, success, operandOrig);
```
Comment (psychocoderHPC) [inline test/unit/atomic/src/AtomicTest.cpp]: Some as above

File: test/unit/atomic/src/AtomicTest.cpp [hunk 3]
```diff
@@ -41,294 +41,663 @@
...
+
+#if BOOST_ARCH_PTX >= BOOST_VERSION_NUMBER(3, 5, 0)
+        testAtomicMin(acc, success, operandOrig);
+        testAtomicMax(acc, success, operandOrig);
+#endif
+
+        testAtomicExch(acc, success, operandOrig);
+
+        //testAtomicInc(acc, success, operandOrig);
```
Comment (psychocoderHPC) [inline test/unit/atomic/src/AtomicTest.cpp]: same.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 4]
```diff
@@ -41,294 +41,663 @@
...
+    //-----------------------------------------------------------------------------
+    ALPAKA_NO_HOST_ACC_WARNING
+    ALPAKA_FN_ACC auto operator()(
+        alpaka::acc::AccGpuCudaRt<TDim, TIdx> const & acc,
+        bool * success,
+        float operandOrig) const
+    -> void
+    {
+        testAtomicAdd(acc, success, operandOrig);
+        //testAtomicSub(acc, success, operandOrig);
```
Comment (psychocoderHPC) [inline test/unit/atomic/src/AtomicTest.cpp]: same

File: test/unit/atomic/src/AtomicTest.cpp [hunk 5]
```diff
@@ -41,294 +41,663 @@
...
+    //-----------------------------------------------------------------------------
+    ALPAKA_NO_HOST_ACC_WARNING
+    ALPAKA_FN_ACC auto operator()(
+        alpaka::acc::AccGpuCudaRt<TDim, TIdx> const & acc,
+        bool * success,
+        double operandOrig) const
+    -> void
+    {
+        testAtomicAdd(acc, success, operandOrig);
+        //testAtomicSub(acc, success, operandOrig);
```
Comment (psychocoderHPC) [inline test/unit/atomic/src/AtomicTest.cpp]: same

File: test/unit/atomic/src/AtomicTest.cpp (hunk 1 above)
Comment (BenjaminW3) [reply test/unit/atomic/src/AtomicTest.cpp]: I have added comments to all those commented out lines, that the operations are not supported.

File: test/unit/atomic/src/AtomicTest.cpp [hunk 6]
```diff
@@ -41,294 +41,677 @@
...
+    TestAtomicOperations<TAcc, unsigned int>::testAtomicOperations();
+    TestAtomicOperations<TAcc, int>::testAtomicOperations();
+#if !(defined(ALPAKA_CI) && BOOST_COMP_MSVC)
+    TestAtomicOperations<TAcc, unsigned long>::testAtomicOperations();
+    TestAtomicOperations<TAcc, long>::testAtomicOperations();
+    TestAtomicOperations<TAcc, unsigned long long>::testAtomicOperations();
+    TestAtomicOperations<TAcc, long long>::testAtomicOperations();
 #endif
     // Not all atomic operations are possible with floating point values.
     //TestAtomicOperations<TAcc, float>::testAtomicOperations();
```
[the remark this answers is not in the document]
Comment (BenjaminW3) [reply test/unit/atomic/src/AtomicTest.cpp]: Yes, with some fixing and/or selection of supported operations this should be possible. The CUDA version should already be functional. The CPU version does not yet work.
I would like to do this in a separate PR after this has been approved and merged.

[the remark this answers is not in the document]
Comment (tdd11235813) [reply test/unit/atomic/src/AtomicTest.cpp]: integrated this PR in HIP PR #505 and all tests passed with HIP(nvcc+hcc).

[the remark this answers is not in the document]
Comment (BenjaminW3) [reply test/unit/atomic/src/AtomicTest.cpp]: Could you please approve this PR?

[the remark this answers is not in the document]
Comment (tdd11235813) [reply test/unit/atomic/src/AtomicTest.cpp]: alright

**Pull request merged**